### PR TITLE
lib: posix: eventfd: general eventfd improvements

### PR DESCRIPTION
**Pull request by e-rk**
The eventfd implementation suffers from various shortcomings.
and it is not thread safe.

This commit addresses the following aspects of eventfd:
* make read and write atomic in respect to each other
* POLLIN after creating eventfd with initval != 0 shall be set
* blocking and nonblocking modes shall have the same effect on poll()
* add support for POLLOUT

Signed-off-by: Rafał Kuźnia <rafal.kuznia@nordicsemi.no>

### CODE REVIEW

File: lib/posix/eventfd.c [hunk 1]
```diff
@@ -49,53 +62,74 @@ static int eventfd_poll_update(struct eventfd *efd,
...
+		key = k_spin_lock(&efd->lock);
+		if ((efd->flags & EFD_NONBLOCK) && efd->cnt == 0) {
+			result = EAGAIN;
+			break;
+		} else if (efd->cnt == 0) {
+			z_pend_curr(&efd->lock, key, &efd->wait_q, K_FOREVER);
+		} else {
+			count = (efd->flags & EFD_SEMAPHORE) ? 1 : efd->cnt;
+			efd->cnt -= count;
+			if (efd->cnt == 0)
```
Comment (andrewboie) [inline lib/posix/eventfd.c:104]: braces

File: lib/posix/eventfd.c [hunk 2]
```diff
@@ -109,22 +143,35 @@ static ssize_t eventfd_write_op(void *obj, const void *buf, size_t sz)
...
+		if ((efd->flags & EFD_NONBLOCK) && overflow) {
+			result = EAGAIN;
+			break;
+		} else if (overflow) {
+			z_pend_curr(&efd->lock, key, &efd->wait_q, K_FOREVER);
 		} else {
-			ret = k_sem_take(&efd->write_sem, K_FOREVER);
+			efd->cnt += count;
+			if (efd->cnt == (UINT64_MAX - 1))
+				k_poll_signal_reset(&efd->write_sig);
```
Comment (andrewboie) [inline lib/posix/eventfd.c:161]: braces

**Review by tyrantlizard77 — COMMENTED** [review]
looks solid just a minor style nit

File: lib/posix/eventfd.c [hunk 3]
```diff
@@ -223,20 +265,29 @@ int eventfd(unsigned int initval, int flags)
...
-	efd->cnt = 0;
-	k_sem_init(&efd->read_sem, 0, 1);
-	k_sem_init(&efd->write_sem, 0, UINT32_MAX);
+	efd->cnt = initval;
+
+	k_poll_signal_init(&efd->write_sig);
+	k_poll_signal_init(&efd->read_sig);
+	z_waitq_init(&efd->wait_q);
+
+	if (initval != 0)
```
[the remark this answers is not in the document]
Comment (e-rk) [reply lib/posix/eventfd.c]: Fixed

**Review by tsvehagen — APPROVED** [review]
LGTM

**Review by mniestroj — APPROVED** [review]

**Pull request merged**